### PR TITLE
LOOP-996: guard against invalid trend (rate) values

### PR DESCRIPTION
**Pull request by rickpasetto**
According to Dexcom:

> The valid range of the rate is [-8.0, 8.0].
>
> If the rate calculated is outside the valid range (it can be positive or negative), then the special value of 12.7 is sent.

This change makes 12.7 (or any other value outside of [-8.0, 8.0]) invalid (and returns NaN).

[LOOP-996](https://tidepool.atlassian.net/browse/LOOP-996)

### CODE REVIEW

File: Models/DexcomCGMReading.swift [hunk 1]
```diff
@@ -10,9 +10,9 @@ import TransmitterCore
 
 struct DexcomCGMReading: TransmitterCgmReading, Codable {
     
-    var recordedSystemTime: Date
+    let recordedSystemTime: Date
```
Comment (rickpasetto) [inline Models/DexcomCGMReading.swift]: Note a small unrelated change: the `TransmitterCgmReading` protocol stipulates that properties only need have getters, so better to make them declared as `let` as much as possible.

File: DexcomCGMKitTests/DexcomCGMManagerTests.swift [hunk 2]
```diff
@@ -304,6 +304,23 @@ class DexcomCGMManagerTests: XCTestCase {
...
+                                                egvSystemTime: expectedDate,
+                                                sequenceNumber: 1,
+                                                algorithmState: AlgorithmState.inCalibration,
+                                                egv: 80,
+                                                predictedEgv: nil,
+                                                rate: 12.7,
+                                                isBgGeneratedOnTransmitter: false,
+                                                aberrationState: AberrationState.none)
+        let cgmReading = DexcomCGMReading(reading)
+        XCTAssertTrue(cgmReading.rate.isNaN)
```
Comment (nhamming) [inline DexcomCGMKitTests/DexcomCGMManagerTests.swift]: Just to make sure we understand the SDK is mappers, could you in `cgmReading.trend` and make sure you get `outOfRange`. Alternatively (or in addition) `transmitterData.trendtype` should return nil.

File: Models/DexcomCGMReading.swift [hunk 3]
```diff
@@ -26,21 +26,32 @@ struct DexcomCGMReading: TransmitterCgmReading, Codable {
...
+    // If the rate calculated is outside the valid range (it can be positive or negative), then the special value of 12.7 is sent.
+    // Given that, we assume _any_ values outside of that range is invalid, and the TransmitterCgmReading protocol says "NaN is used for invalid".
+    let validRateRange = -8.0...8.0
+    var rate: Double {
+        didSet {
+            if !validRateRange.contains(rate) {
+                rate = Double.nan
+            }
+        }
+    }
```
Comment (nhamming) [inline Models/DexcomCGMReading.swift]: Following your update, rate is defined as `var rate: Double { get }`. Could `let` be used here and only set in the `init(_)`?

Comment (rickpasetto) [reply Models/DexcomCGMReading.swift]: I tried that at first (it was my preferred implementation), but I thought I found code paths that modified it after the fact.  Now that I look again, I can't find them.  I'll put it back, but I swore I had trouble with it...

File: DexcomCGMKitTests/DexcomCGMManagerTests.swift (hunk 2 above)
Comment (rickpasetto) [reply DexcomCGMKitTests/DexcomCGMManagerTests.swift]: I did this (sorry, I suppose I should also add information about what I _didn't_ put into the PR as well...).  There's already a unit test, `testTrendTypeOutOfRange`, that does this.  I also confirmed in the UI that `nil` is properly treated and shown in the UI (`TransmitterCgmReading.trendRate`, tested in `testTrendRateInvalid`).  You have good unit tests!

Comment (nhamming) [reply DexcomCGMKitTests/DexcomCGMManagerTests.swift]: :) sorry forgot about that.

Comment (nhamming) [reply DexcomCGMKitTests/DexcomCGMManagerTests.swift]: :) sorry I forgot about that.
